ddl/ddl_tiflash_test.go

```diff
@@ -649,3 +667,170 @@ func TestAlterDatabaseBasic(t *testing.T) {
 	// There is less TiFlash store
 	tk.MustGetErrMsg("alter database tiflash_ddl set tiflash replica 3", "the tiflash replica count: 3 should be less than the total tiflash server count: 2")
 }
+
+func checkBatchPandingNum(t *testing.T, tkx *testkit.TestKit, level string, value string, ok bool) {
+	l := len(tkx.MustQuery(fmt.Sprintf("show %v variables where Variable_name='tidb_batch_pending_tiflash_count' and Value='%v'", level, value)).Rows())
+	if ok {
+		require.Equal(t, 1, l)
+	} else {
+		require.Equal(t, 0, l)
+	}
+}
+
+func TestTiFlashBatchAddVariables(t *testing.T) {
+	store, _, tear := createStoreWithoutMockTiFlash(t)
+	defer tear()
+
+	tk := testkit.NewTestKit(t, *store)
+	tk.MustExec("set SESSION tidb_batch_pending_tiflash_count=5")
+	tk.MustExec("set GLOBAL tidb_batch_pending_tiflash_count=6")
+
+	checkBatchPandingNum(t, tk, "session", "5", true)
+	checkBatchPandingNum(t, tk, "global", "6", true)
+	checkBatchPandingNum(t, tk, "global", "1.5", false)
+
+	tk.MustGetErrMsg("set GLOBAL tidb_batch_pending_tiflash_count=1.5", "[variable:1232]Incorrect argument type to variable 'tidb_batch_pending_tiflash_count'")
+	checkBatchPandingNum(t, tk, "global", "6", true)
+
+	tk2 := testkit.NewTestKit(t, *store)
+	checkBatchPandingNum(t, tk2, "session", "6", true)
+}
+
+func execWithTimeout(t *testing.T, tk *testkit.TestKit, to time.Duration, sql string) (bool, error) {
+	timeOut := atomicutil.NewBool(false)
+	ctx, cancel := context.WithTimeout(context.Background(), to)
+	defer cancel()
+	enabled := atomicutil.NewBool(false)
+	doneCh := make(chan int, 1)
+	var wg sync.WaitGroup
+	go func() {
+		defer wg.Done()
+		select {
+		case <-doneCh:
+			// Exit normally
+			return
+		case <-ctx.Done():
+			// Exceed given timeout
+		}
+		logutil.BgLogger().Info("execWithTimeout meet timeout", zap.String("sql", sql))
+		timeOut.Store(true)
+		enabled.Store(true)
+		require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/ddl/BatchAddTiFlashSendDone", "return(true)"))
```

Comment:
I suggest you swap the go routine and exec, I.e. `tk.exec + doneCh <-` in goroutine, and use select to wait outside `select { doneCh ctx.Done }`.

Then you could also remove the wait group, since you could `<- doneCh` to wait for the exiting. Enabled could also be removed: `failpoint.enable; wait tk.exec; failpoint.disable`.